### PR TITLE
Convert data for all Genie Engine games

### PR DESCRIPTION
**Pull request by heinezen**
Aims to add support for all Genie Engine games to the converter.

- [x] More versatile converter
- [x] Age of Empires I
- [x] Rise of Rome
- [x] Age of Kings
- [x] The Conquerors + African Kingdoms (full support)
- [x] Star Wars: Galactic Battlegrounds
- [x] Star Wars: Galactic Battlegrounds: Clone Campaigns

### CODE REVIEW

**Comment by heinezen** [comment]
@TheJJ I want to keep it granular on a granular level for now because different Genie versions require different data fields to be added and read pretty much everywhere in the `.dat` file. Once all the games are covered by the converter, optimization will start ;-)

**Comment by heinezen** [comment]
Ok, this is done now. I updated everything I could, except for two that are supposed to be removable, but break our conversion script. Everything related to SWGB and AOE1 is commented out because I would have to change too much of the converter for it to be useful right now. Once we have figured out the `nyan` integration, we'll have to rewrite parts of the converter anyway.

**Comment by TheJJ** [comment]
We should create the data format at runtime with objects, not with classes.
We used the class-approach because "the buildsystem needs to generate headers at compiletime".
With `nyan`, the convert export format can be done at runtime (i.e. that SWGB creates files for the same engine api as AoC uses it).

File: openage/convert/gamedata/maps.py [hunk 1]
```diff
@@ -86,39 +92,44 @@ class MapUnit(Exportable):
...
-    )
+    ]
 
 
-class MapUnknown(Exportable):
+class MapElevation(Exportable):
     name_struct_file   = "randommap"
-    name_struct        = "map_unknown"
+    name_struct        = "map_elevation"
     struct_description = "random map infos about magic unknown things"
```
Comment (TheJJ) [inline openage/convert/gamedata/maps.py]: pls update the description

File: openage/convert/gamedata/tech.py [hunk 2]
```diff
@@ -188,30 +280,49 @@ class UnitConnection(Exportable):
...
-        (READ, "required_researches", "int32_t"),
-        (READ, "age", "int32_t"),
-        (READ, "unit_or_research0", "int32_t"),
-        (READ, "unit_or_research1", "int32_t"),
-        (READ_UNKNOWN, None, "int32_t[8]"),
-        (READ, "mode0", "int32_t"),
-        (READ, "mode1", "int32_t"),
-        (READ_UNKNOWN, None, "int32_t[7]"),
+
+        (READ_EXPORT, "slots_used", "int32_t"),
```
Comment (TheJJ) [inline openage/convert/gamedata/tech.py]: why export?

File: openage/convert/gamedata/unit.py [hunk 3]
```diff
@@ -523,24 +537,43 @@ class UnitObject(Exportable):
...
+    # if (GameVersion.aoe_1 or GameVersion.aoe_ror) not in game_versions:
+    #     data_format.append((READ_EXPORT, "damage_sound", "int16_t"))
+    # ===========================================================================
+    data_format.append((READ_EXPORT, "damage_sound", "int16_t"))
+
+    data_format.extend([
+        (READ_EXPORT, "dead_unit_id", "int16_t"),        # unit id to become on death
         (READ, "placement_mode", "int8_t"),              # 0=placable on top of others in scenario editor, 5=can't
-        (READ, "air_mode", "int8_t"),                    # 1=no footprints
+        (READ, "can_be_nuilt_on", "int8_t"),             # 1=no footprints
```
Comment (TheJJ) [inline openage/convert/gamedata/unit.py]: typo?

File: openage/convert/gamedata/unit.py [hunk 4]
```diff
@@ -719,9 +813,9 @@ class UnitObject(Exportable):
             },
         )),
         (READ, "editor_selection_color", "uint8_t"),  # 0: default, -16: fish trap, farm, 52: deadfarm, OLD-*, 116: flare, whale, dolphin -123: fish
-        (READ_EXPORT, "selection_shape_x", "float"),
-        (READ_EXPORT, "selection_shape_y", "float"),
-        (READ_EXPORT, "selection_shape_z", "float"),
+        (READ_EXPORT, "outline_size_x", "float"),
+        (READ_EXPORT, "outline_size_y", "float"),
+        (READ_EXPORT, "outline_size_z", "float"),
```
Comment (TheJJ) [inline openage/convert/gamedata/unit.py]: i think this is the circle or square when a unit is selected, not the outline (which is the pixel-correct background-unit border).

File: openage/convert/gamedata/unit.py [hunk 5]
```diff
@@ -1030,24 +1309,11 @@ class BuildingUnit(LivingUnit):
             },
         )),
         (READ, "garrison_heal_rate", "float"),
-        (READ_UNKNOWN, None, "float"),      # (unknown garrison value)
+        (READ, "garrison_repair_rate", "float"),              # (unknown garrison value)
```
Comment (TheJJ) [inline openage/convert/gamedata/unit.py]: comment update pls

File: openage/convert/gamedata/tech.py [hunk 6]
```diff
@@ -221,19 +332,48 @@ class ResearchConnection(Exportable):
...
-        (READ, "age", "int32_t"),
-        (READ, "upper_research", "int32_t"),
-        (READ_UNKNOWN, None, "int32_t[9]"),
-        (READ, "line_mode", "int32_t"),
-        (READ_UNKNOWN, None, "int32_t[8]"),
+    ])
+    # ===========================================================================
+
+    data_format.extend([
+        (READ_EXPORT, "slots_used", "int32_t"),
```
Comment (TheJJ) [inline openage/convert/gamedata/tech.py]: why export?

**Comment by heinezen** [comment]
@TheJJ Should be fixed now.

**Pull request merged**